include/vinecopulib/bicop/implementation/parametric.ipp

```diff
@@ -56,126 +56,126 @@ inline void ParBicop::fit(const Eigen::Matrix<double, Eigen::Dynamic, 2> &data,
-        return;
-    }
+      // for independence copula we don't have to do anything
+      if (family_ == BicopFamily::indep) {
+          set_loglik(0.0);
+          return;
+      }
 
-    using namespace tools_optimization;
-    std::vector <std::string> methods = {"itau", "mle"};
-    if (!tools_stl::is_member(method, methods)) {
-        throw std::runtime_error("Method not implemented.");
+      check_fit_method(method);
+      double tau = wdm::wdm(data, "tau", weights)(0, 1);
+
+      // for method itau and one-parameter families we don't need to optimize
+      int npars = static_cast<int>(calculate_npars()) - (method == "itau");
+      if (npars == 0) {
+          set_parameters(tau_to_parameters(tau));
+          set_loglik(loglik(data, weights));
+          return;
+      }
+
+      // Set bounds and starting values
+      auto lb = get_parameters_lower_bounds();
+      auto ub = get_parameters_upper_bounds();
+      adjust_parameters_bounds(lb, ub, tau, method);
+      auto initial_parameters = get_start_parameters(winsorize_tau(tau));
+
+      // find (pseudo-) mle
+      std::function<double(const Eigen::VectorXd&)> objective;
+      if (method == "mle") {
+          objective = [&data, &weights, this] (const Eigen::VectorXd& pars) {
+              this->set_parameters(pars);
+              return this->loglik(data, weights);
+          };
+      } else {
+          // profile likelihood
+          set_parameters(initial_parameters);
+          initial_parameters(0) = initial_parameters(1);
+          initial_parameters.conservativeResize(1);
+          objective = [&data, &weights, this] (const Eigen::VectorXd& pars) {
+              Eigen::VectorXd newpars(2);
+              newpars(0) = this->get_parameters()(0);
+              newpars(1) = pars(0);
+              this->set_parameters(newpars);
+              return this->loglik(data, weights);
+          };
+      }
+
+      tools_optimization::Optimizer optimizer;
+      auto newpars = optimizer.optimize(initial_parameters, lb, ub, objective);
+
+      // check if fit is reasonable, otherwise increase search interval
+      // and refit
+      if (optimizer.get_objective_max() < -0.1) {
+          newpars = optimizer.optimize(initial_parameters,
+                                       get_parameters_lower_bounds(),
+                                       get_parameters_upper_bounds(),
+                                       objective);
+      }
+
+      // finalize fitted model
+      if (method == "itau") {
+          // only the second parameter has been optimized
+          newpars.conservativeResize(2);
+          std::swap(newpars(0), newpars(1));
+          newpars(0) = get_parameters()(0);
+      }
+
+      set_parameters(newpars);
+      set_loglik(optimizer.get_objective_max());
+}
+
+//! ensures that starting values are sufficiently separated from bounds
+//! @param tau kendall's tau
+inline double ParBicop::winsorize_tau(double tau) const
```

Comment:
Funny name, winsorize would be more when tau is close to 0 or 1 no?

Reply:
Well it does enforce that |tau| is in [0.01, 0.9], but I'm open to other suggestions.

Reply:
I meant -1 and 1, my bad, and I do totally understand the logic, I just wondered about whether a better name could be found (winsorization usually doesn't do anything in the middle of the interval). In any case, I'm fine with this.